macsio/macsio_main.c

```diff
@@ -779,6 +781,36 @@ main_read(int argi, int argc, char **argv, json_object *main_obj)
     return (0);
 }
 
+static void InitializePRNGs(void)
+{
+    double currtime = MT_Time();
+    unsigned ucurrtim = ((unsigned *)&currtime)[0] ^ ((unsigned *)&currtime)[1];
+
+    /* ensure all procs have same ucurrtim */
+#ifdef HAVE_MPI
+    MPI_Bcast(&ucurrtim, 1, MPI_UNSIGNED, 0, MACSIO_MAIN_Comm);
+#endif
+
+    /* Initialize the PRNGs */
+    MACSIO_DATA_InitializePRNGs((unsigned) MACSIO_MAIN_Rank, ucurrtim);
+}
+
+static void FinalizePRNGs()
+{
+#ifdef HAVE_MPI
+    /* Lets confirm the rank-invariant PRNGs agree and issue a message if not */
+    unsigned rand_check[2] = {MD_random_rankinv(), MD_random_rankinv_tv()};
```

Comment:
How can you ensure that the time-variant random is identical across all ranks? What ensures that the time is identical?

Reply:
The `_rankinv` PRNGs are identical across all ranks because they are all seeded by the same seed. The `_tv` variant uses a seed that varies with time but is, nonetheless, identical across all ranks by an additional broadcast in the initialization phase. Does that make sense?

Reply:
Also, we can catch incorrect usage of the `_rankinv` PRNGs at termination. They should all agree across ranks and if they do not, someone, somewhere, has used a `_rankinv` PRNG in a non-collective way.